File: RealtimeTranslator/OpenAI/TranslationPumpSupervisor.swift
import Foundation

/// 翻訳送信ポンプの task ハンドル・世代番号・transport halt・連続失敗数を管理する。
/// ポンプ本体のループはオーケストレーターが持ち、ここは帳簿だけを担う。
/// `DualRealtimeTranslationClient` の actor 隔離下だけで触る値型コンポーネント。
struct TranslationPumpSupervisor {
    private(set) var task: Task<Void, Never>?
    /// 現在登録中のポンプ世代。古いポンプの終了処理が新ポンプの参照を消さない。
    private(set) var generation = 0
    /// transport failure後、再接続まで翻訳ポンプを再開しない。
    private(set) var haltedForTransportFailure = false
    private(set) var consecutiveFailures = 0
    private let failureLimit: Int

    init(failureLimit: Int) {
        self.failureLimit = failureLimit
    }

    var isTracked: Bool {
        task != nil
    }

    /// 新しい世代としてポンプを登録する。task が既にある場合のガードは呼び出し側。
    mutating func start(body: @escaping @Sendable (Int) async -> Void) {
        generation += 1
        let current = generation
        task = Task {
            await body(current)
        }
    }

    /// 現在世代なら task を手放す。再開可否の判定は呼び出し側が行う。
    mutating func finishIfCurrent(generation expected: Int) -> Bool {
        guard generation == expected else { return false }
        task = nil
        return true
    }

    mutating func invalidate() {
        generation += 1
        task?.cancel()
        task = nil
    }

    mutating func haltForTransportFailure() {
        haltedForTransportFailure = true
    }

    @discardableResult
    mutating func recordFailure() -> Int {
        consecutiveFailures += 1
        return consecutiveFailures
    }

    var reachedFailureLimit: Bool {
        consecutiveFailures >= failureLimit
    }

    mutating func resetFailures() {
        consecutiveFailures = 0
    }

    mutating func reset() {
        haltedForTransportFailure = false
        consecutiveFailures = 0
    }
}

File: windows/src/RealtimeTranslator.Core/Realtime/TranslationPumpSupervisor.cs
using System;
using System.Threading;
using System.Threading.Tasks;

namespace RealtimeTranslator.Core.Realtime;

/// <summary>
/// 翻訳送信ポンプの task ハンドル・世代番号・transport halt・連続失敗数を管理する。
/// ポンプ本体のループはオーケストレーターが持ち、ここは帳簿だけを担う。
/// `DualRealtimeTranslationClient` の `_sync` ロック下だけで触るコンポーネント。
/// </summary>
internal sealed class TranslationPumpSupervisor
{
    private readonly int _failureLimit;

    public TranslationPumpSupervisor(DualRealtimeTranslationClientTuning tuning)
    {
        ArgumentNullException.ThrowIfNull(tuning);
        _failureLimit = tuning.ConsecutiveFailureLimit;
    }

    public Task? PumpTask { get; private set; }

    /// <summary>現在登録中のポンプ世代。古いポンプの終了処理が新ポンプの参照を消さない。</summary>
    public int Generation { get; private set; }

    /// <summary>transport failure 後は再接続まで翻訳ポンプを再開しない。</summary>
    public bool HaltedForTransportFailure { get; private set; }

    public int ConsecutiveFailures { get; private set; }

    public CancellationTokenSource Cancellation { get; private set; } = new();

    public bool IsTracked => PumpTask is not null;

    public bool ReachedFailureLimit => ConsecutiveFailures >= _failureLimit;

    /// <summary>新しい世代としてポンプを登録する。task が既にある場合のガードは呼び出し側。</summary>
    public void Start(Task task)
    {
        Generation += 1;
        PumpTask = task;
    }

    /// <summary>現在世代なら task を手放す。再開可否の判定は呼び出し側が行う。</summary>
    public bool FinishIfCurrent(int generation)
    {
        if (Generation != generation)
        {
            return false;
        }

        PumpTask = null;
        return true;
    }

    /// <summary>ポンプ参照を切り離して返す。cancel / await はロック外で呼び出し側が行う。</summary>
    public (Task? PumpTask, CancellationTokenSource Cancellation) Detach()
    {
        var task = PumpTask;
        var cts = Cancellation;
        Generation += 1;
        PumpTask = null;
        return (task, cts);
    }

    /// <summary>次の接続世代向けにキャンセル源を張り直す。</summary>
    public void RecycleCancellation()
    {
        Cancellation.Dispose();
        Cancellation = new CancellationTokenSource();
    }

    public void HaltForTransportFailure() => HaltedForTransportFailure = true;

    public int RecordFailure() => ++ConsecutiveFailures;

    public void ResetFailures() => ConsecutiveFailures = 0;

    public void Reset()
    {
        HaltedForTransportFailure = false;
        ConsecutiveFailures = 0;
    }
}
